### DISCUSSION TITLE
IRdfHandlers for writing triples in a streaming fashion

### DISCUSSION THREAD

**Discussion by IS4Code** [dotnetrdf/dotnetrdf#564] [Ideas]
Hello, I was wondering what your opinion on having implementations of `IRdfHandler` that actually write the RDF data to a stream in a specific serialization format. Since I had a need for something like this, and because I wanted a bit different style, I made my own:

<details>
  <summary><code>TurtleHandler</code> (pre-3.0)</summary>
  
  ```js
  using System;
  using System.Collections.Generic;
  using System.IO;
  using VDS.RDF;
  using VDS.RDF.Parsing.Handlers;
  using VDS.RDF.Writing;
  using VDS.RDF.Writing.Formatting;

  /// <summary>
  /// A custom RDF handler that writes Turtle output to a file.
  /// </summary>
  /// <typeparam name="TFormatter">
  /// The Turtle formatter used for nodes, usually <see cref="TurtleFormatter"/>.
  /// </typeparam>
  public class TurtleHandler<TFormatter> : BaseRdfHandler where TFormatter : INodeFormatter, ITripleFormatter, IUriFormatter, INamespaceFormatter, IBaseUriFormatter
  {
      readonly TextWriter output;
      readonly TFormatter formatter;
      readonly INamespaceMapper namespaceMapper;
      readonly IEqualityComparer<Uri> uriComparer;

      string? lastSubject;
      Uri? lastBase;

      /// <summary>
      /// Creates a new instance of the Turtle handler.
      /// </summary>
      /// <param name="output">The output writer to write Turtle to.</param>
      /// <param name="formatter">The formatter of Turtle nodes.</param>
      /// <param name="namespaceMapper">The namespace mapper to register namespaces to.</param>
      /// <param name="uriComparer">The comparer of <see cref="Uri"/> instances.</param>
      public TurtleHandler(TextWriter output, TFormatter formatter, INamespaceMapper namespaceMapper, IEqualityComparer<Uri>? uriComparer = null)
      {
          this.output = output;
          this.formatter = formatter;
          this.namespaceMapper = namespaceMapper;
          this.uriComparer = uriComparer ?? new UriComparer();
      }

      /// <inheritdoc/>
      public override bool AcceptsAll => true;

      private void EndSubject(string? test)
      {
          if(lastSubject != test && lastSubject != null)
          {
              if(lastSubject != "@base" && lastSubject != "@prefix")
              {
                  WriteLine(" .");
              }
              WriteLine();
          }
          lastSubject = test;
      }

      /// <inheritdoc/>
      protected override bool HandleNamespaceInternal(string prefix, Uri namespaceUri)
      {
          EndSubject("@prefix");
          WriteLine(formatter.FormatNamespace(prefix, ResolveUri(namespaceUri)));
          namespaceMapper.AddNamespace(prefix, namespaceUri);
          return true;
      }

      /// <inheritdoc/>
      protected override bool HandleBaseUriInternal(Uri baseUri)
      {
          if(lastBase == null || !uriComparer.Equals(baseUri, lastBase))
          {
              EndSubject("@base");
              WriteLine(formatter.FormatBaseUri(ResolveUri(baseUri)));
              lastBase = baseUri;
          }
          return true;
      }

      /// <inheritdoc/>
      protected override void EndRdfInternal(bool ok)
      {
          base.EndRdfInternal(ok);
          EndSubject(null);
          Flush();
      }

      private Uri ResolveUri(Uri uri)
      {
          if(lastBase != null)
          {
              var relative = lastBase.MakeRelativeUri(uri);
              if(!relative.IsAbsoluteUri && uriComparer.Equals(uri, new Uri(lastBase, relative)) && relative.OriginalString.Length < uri.AbsoluteUri.Length)
              {
                  return new FakeUri(uri.AbsoluteUri, relative.OriginalString);
              }
          }
          return uri;
      }

      private INode ResolveNode(INode node)
      {
          if(lastBase != null && node is IUriNode uriNode)
          {
              return CreateUriNode(ResolveUri(uriNode.Uri));
          }
          return node;
      }

      /// <inheritdoc/>
      protected override bool HandleTripleInternal(Triple t)
      {
          var subj = formatter.Format(ResolveNode(t.Subject), TripleSegment.Subject).Trim();
          var pred = formatter.Format(ResolveNode(t.Predicate), TripleSegment.Predicate).Trim();
          var obj = formatter.Format(ResolveNode(t.Object), TripleSegment.Object).Trim();
          if(lastSubject != subj)
          {
              EndSubject(null);
              if(pred == "a")
              {
                  Write($"{subj} a {obj}");
              }else{
                  WriteLine(subj);
                  Write($"  {pred} {obj}");
              }
              lastSubject = subj;
          }else{
              WriteLine(" ;");
              Write($"  {pred} {obj}");
          }
          return true;
      }

      string linePart = "";
      void Write(string str)
      {
          linePart += str;
      }

      void WriteLine(string? str = null)
      {
          output.WriteLine(linePart + str);
          linePart = "";
      }

      void Flush()
      {
          output.Write(linePart);
          output.Flush();
          linePart = "";
      }

      /// <summary>
      /// A fake <see cref="Uri"/> that has a custom <see cref="ToString"/> value,
      /// to give to the <typeparamref name="TFormatter"/>.
      /// </summary>
      class FakeUri : Uri
      {
          readonly string stringValue;

          public FakeUri(string realValue, string stringValue) : base(realValue)
          {
              this.stringValue = stringValue;
          }

          /// <inheritdoc/>
          public override string ToString()
          {
              return stringValue;
          }
      }
  }
  ```
</details>

I am in a situation where triples are produced during the whole operation of the software, and I feel there are not enough options for serializing RDF data in this case. As far as I am aware, there are two possibilities:

* Store all triples in a graph and then pick a `IRdfWriter` implementation to serialize it all at once. This can be generalized thanks to `MimeTypeDefinition.GetRdfWriter()`, but does not work in a streaming way.
* Find a `ITripleFormatter` and use it together with `WriteThroughHandler`. It is possible to obtain the formatter from `IFormatterBasedWriter`, but since all a formatter sees are individual triples, this method cannot offer any compact syntax for groups of triples (unlike my `TurtleHandler` which remembers the subject). It also does not work in all situations (for example the formatter for CSV outputs space-separated nodes instead of comma-separated).

What I think would be nice is to have something in between those two options, and I think `IRdfHandler` suits it well. Preferably, each `MimeTypeDefinition` should expose the underlying formatter (if suitable for triples), but also a way to create `IRdfHandler` to `TextWriter` which could do more fancy stuff, and perhaps support other things like `IPrettyPrintingWriter`.

I am also planning to write something like this for JSON-LD, which is a bit wild, but we shall see if it works.

**Comment by kal** [comment]
I can see how this would be useful for implementing streaming output from a process that is generating triples. The simple `IRdfHandler` interface is a reasonable start, but I think you would probably want to add some additional methods such as marking the start and end of graphs, maybe also allowing comments to be inserted into the stream. There's also the question of how chaining of triples is handled where you might have multiple statements with the same subject, or same subject + predicate - that could be done by a single `Triple(subj, pred, obj)` method that then puts the responsibility on the writer to emit triples in order, or it could use a more XML/SAX style approach with `StartSubject(s)`/`EndSubject()`, `StartPredicate(p)`/`EndPredicate()` and `Object(o)` methods that would let you chain without having to constantly repeat the unchanged elements of each triple. Not sure if that would be easier or harder overall, but its just another thing to maybe consider in the design stage.

 There's 12 formats for RDF that are currently supported by the library (that's not including the GZip variants or any of the SPARQL results formats) so implementing this for all of the formats (and writing the necessary unit tests and documentation) is not a small task (although it should be possible to reuse a lot of the logic from the existing writers as well as make use of the existing formatters)

**Reply by IS4Code** [reply]
Starting with the comments, I think the `ICommentRdfHandler` interface I added covers that sufficiently well. ☺ You are right that having a greater degree of control over various "sections" that the serialization formats offer (including multi-line comments) might be advantageous, but I am wondering if the complexity this would bring is worth it, especially considering that not all formats have the same capabilities.

However, having tried to implement the handler for JSON-LD, I have faced situations where the specific heuristics or strategy of opening or closing a particular "context" might be useful to decouple. For example, you could have several nested objects, each having several properties, and all linked to the parent object via some property:

```json
{
  "ex:prop": [
    { "ex:prop2": 1 },
    { "ex:prop2": 1 },
    { "ex:prop2": 1 }
  ]
}
```

My current implementation only "buffers" at most one triple (it waits before writing the object), but in this situation, this would be emitted using `IRdfHandler` as 6 interspersed triples, and the handler will not know that there are multiple values of `ex:prop`, which is a catastrophe in JSON, turning this into:

```json
{
  "ex:prop": { "ex:prop2": 1 },
  "@nest": {
    "ex:prop": { "ex:prop2": 1 },
    "@nest": {
      "ex:prop": { "ex:prop2": 1 }
    }
  }
}
```

Thankfully I can fix this in a better way, by using new unique prefixes:

```json
{
  "ex:prop": { "ex:prop2": 1 },
  "ex|2:prop": { "ex:prop2": 1 },
  "ex|3:prop": { "ex:prop2": 1 }
}
```

But if the handler could receive a hint that multiple values will come (by means of `StartPredicate`), it could start with the array straight away.

Similarly, in Turtle, there is a potential of nesting blank nodes, but only if the user promised that the node could remain unlabelled somehow. Then Turtle could do nesting and JSON-LD or RDF/XML could omit `@id`/``rdf:nodeId`, practically in the same situations.

**Reply by kal** [reply]
Oh yes, I had forgotten about `ICommentRdfHandler`. I think for multi-line comments we could document in the interface that a handler SHOULD split the input string on new lines and either emit multiple single-line comments or a single mult-line comment, but equally leave it open for handlers to simply escape newlines as they might other characters not allowed in a single-line comment.

I agree that the chaining of contexts is the biggest issue for implementers who want to make use of the various compact representations allowed in different syntaxes. On the other hand it also complicates the invocation for clients that just want to stream out a bunch of triples as NTriples or as only lightly compressed turtle (with prefixes and limited nesting but no compact lists for example), and also complicates the implementation for syntaxes like NTriples where there is no sort of compaction.

Maybe there just needs to be multiple forms of sink interface that we provide:

- Simple triples + optional comments + optional start/end graph
- Chained RDF

Adapters could provide a way for writers to use a complex chained RDF instance as if it were a simple triples instance, and an abstract class could provide the state management functionality to make it easier to implement chained RDF handlers (keeping track of the current chaining state).

Alternatively we could just define the chained RDF interface as being the interface that all sinks are expected to implement and provide that adapter on top as an implementation of `IRdfHandler` + `ICommentRdfHandler`.

What do you think? I guess that other than NTriples/NQuads all of the syntaxes would benefit from the chained approach so maybe that should be the default interface?

**Reply by IS4Code** [reply]
Indeed, I think there should definitely be an adapter that just implements `IRdfHandler`/`ICommentRdfHandler` and picks a particular strategy to achieve the desired complexity and compactness of the output, chained to "dumb" format-specific handlers that just write what is given to them.

The rest depends on what kinds of features should be supported:

* Sequential triple building (such as unlabeled blank nodes/lists in Turtle in subject and object, i.e. `[ ... ] ex:prop [ ... ] .`; `rdf:parseType="Resource"` in RDF/XML).
* Blank node property list (only Turtle, not meaningful elsewhere).
* Grouped by subject (Turtle; objects in JSON-LD; `rdf:Description` in RDF/XML).
* Grouped by subject and predicate (Turtle; property array values in JSON-LD; omit `rdf:resource` in RDF/XML).
* Collections. (Turtle; `@list` in JSON-LD; `rdf:parseType="Collection"` in RDF/XML).
* Container membership. (`rdf:li` in RDF/XML).
* Native literals (`@json`, `rdf:parseType="Literal"`).

Ideally, if all of this is covered, even the existing graph or dataset-based writers could be decoupled to use these RDF builders together with a strategy to select and order the appropriate triples.

Even when using just plain old `IRdfHandler`, there are at least two things that should be controlled by the user:

* A desire whether a particular node could be "compacted", for example selecting between `true` and `"true"^^xsd:boolean`, `()` and `rdf:nil`, or using native literals when possible. This might be represented by this sort of interface:
  ```cs
  public interface ICompactedNode : INode
  {
      bool? ShouldCompact { get; }
  }
  ```
  Expressive desire whether the value should or should not be compacted. The default depends on the kind of value ‒ I think `true` should be preferred over `"true"^^xsd:boolean` and `()` over `rdf:nil`, but in situations where the literal value of the node could be lost (for example for native literals or unusually formatted numeric values), they should not be compacted by default (for example `"001"^^xsd:integer` should stay that way unless `ShouldCompact == true`, turning it to `1`).
* A temporary unlabeled blank node and an indication of its lifetime, so that the builder may know whether it's safe to omit its identifier (use `[`,`]` in Turtle or skip `@id`/`rdf:nodeID`). This could be governed by this interface:
  ```cs
  public interface ITemporaryBlankNode : IBlankNode, ICompactedNode, IDisposable
  {
      bool Disposed { get; }
  }
  ```
  Such a blank node would have effectively meaningless identifier and instead be compared purely by identity, plus once disposed, using it would not be allowed. A particular implementation may buffer triples using this kind of node until `Disposed == true`, and then write it all at once with the knowledge that it will never be used again.
  
  Additionally, if `ShouldCompact == true` here, the builder may take it as the indication that the node may be unlabeled straight away, such as by beginning to write `[` in Turtle. Once the corresponding `]` is written and it is no longer possible to refer to the node, the builder could itself call `Dispose`, preventing the user from making the mistake of using it again.

It should be noted that I did not take named graphs and RDF-star much into account here (since I haven't used them extensively yet), so these concepts may be adjusted to accommodate for them too.

Lastly, of all the 3 formats I have used as examples, RDF/XML is arguably the most complex one, with many shorthands that result in more compact output, such as replacing `rdf:Description` with a concrete type, property attributes for string literals and `rdf:type`, `rdf:parseType="Resource"` etc. It would be great if whatever builder interface is come up with here could map into those features optimally.

**Reply by kal** [reply]
The issue I have with extending the `INode` interface is that you are then requiring the caller to create node instances that implement those interfaces which might be an issue if the nodes are coming from some upstream process such as a SPARQL query results set. So I wonder if there are some alternative appraoches that don't require that. 

With regards to compacting object node values, I think a simple boolean flag on the Object() method would be sufficient and should maybe be a `noCompact` flag that would essentially disable whatever the custom compression level is set to for the handler. 

For unlabelled nodes I think there's a bunch of options with different tradeoffs. Maybe the simplest would be to have a `StartSubject` overload that takes no arguments - this is essentially saying "serialize a blank node, which can be unlabelled". Another would be to have a flag on `StartSubject` that indicates if the label of a blank node should be retained. I think I prefer the first of these two options.

For containers and lists I reckon there are maybe two approaches to go with this. The first is to leave it all up to the handler and document a convention (e.g. say that handlers expect `<foo> a rdf:List` before any other triples with `<foo>` as a subject if you want them to handle `<foo>` as a list, similarly with the collection types). The second would be to have explicit support with `StartCollection`/`EndCollection` and `StartContainer`/`EndContainer` methods. I think that the second approach is the one I would prefer...it does imply a more complex interface to be implemented, but it maybe is the means that opens up most options in terms of supporting the syntactic sugar for collections that are provided by the various syntaxes.

I wonder if any other compression options could then be handled by configuration options passed to the handler when it is constructed rather than having more methods or arguments on the basic chained RDF interface.

**Reply by kal** [reply]
Here's a scrappy at outlining what the interface might be like:
https://gist.github.com/kal/b21e3b4350c19994820f52d55cec9237

The intended use would be something like:

```csharp
handler.StartSubject(s);
handler.StartPredicate(p);
handler.Object(o1); // can emit (s, p, o1)
handler.Object(o2); // can emit (s, p, o2);
handler.EndPredicate();
handler.StartPredicate(p2);
handler.StartCollection(); // value of p2 is an anonymous blank node that is an rdf:List
handler.Object(m1); // First list item (with an anonymous blank node for the list member)
handler.StartSubject(m2); // Second list item (with an anonymous blank node for the list member)
handler.Predicate(p3); // Second list item has a property that can be serialised inline
handler.Object(o3); 
handler.EndSubject(); // End m2
handler.StartMember(listNode); // Just for fun lets have a named list node
handler.Object(o4); // The rdf:first value for listNode
handler.EndMember(); // Pop the list node
handler.EndCollection(); // End of list
handler.EndPredicate(); // End of p2
handler.EndSubject(); // End of s
```

**Reply by kal** [reply]
Hmm...I wonder if the methods for collections should be better aligned to the RDF Model so we should maybe be able to write code like:

```
handler.StartCollection();
handler.StartMember();
handler.StartFirst();
// code to emit first list item
handler.EndFirst();
handler.StartRest();
// code to emit other list items
handler.EndRest();
handler.EndMember();
handler.EndCollection();
```

Its quite verbose, but you could provide overloads like handler.First(node); and maybe say that if handler.First or handler.StartFirst are called without a preceding call to handler.StartMember, then an anonymous blank node is generated for the list member.
